src/lib/support/SerializableIntegerSet.cpp

```diff
@@ -0,0 +1,132 @@
+
+    buflen = Base64Encode(reinterpret_cast<const uint8_t *>(mData), static_cast<uint16_t>(len), buf);
+    out    = buf;
+
+exit:
+    return out;
+}
+
+CHIP_ERROR SerializableU64SetBase::DeserializeBase64(const char * serialized, uint16_t buflen)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    uint16_t decodelen = 0;
+    VerifyOrExit(buflen <= MaxSerializedSize(), err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    decodelen = Base64Decode(serialized, buflen, reinterpret_cast<uint8_t *>(mData));
+    VerifyOrExit(decodelen <= sizeof(uint64_t) * mCapacity, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(decodelen % sizeof(uint64_t) == 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    mNextAvailable = decodelen / static_cast<uint16_t>(sizeof(uint64_t));
+exit:
+    return err;
+}
+
+uint16_t SerializableU64SetBase::Find(uint64_t value)
+{
+    /**
+     *  The data is stored in LittleEndian byte order in the set. This will enable
+     *  different machine architectures to interpret a given set in a consistent manner,
+     *  for serialize and deserialize operations.
+     */
+    uint64_t valueLE = ToLE64(value);
+    for (uint16_t i = 0; i < mNextAvailable; i++)
+    {
+        if (mData[i] == valueLE)
+        {
+            return i;
+        }
+    }
+
+    return mCapacity;
+}
+
+CHIP_ERROR SerializableU64SetBase::Insert(uint64_t value)
+{
+    CHIP_ERROR err = CHIP_ERROR_NO_MEMORY;
+    uint16_t index = FirstAvailableForUniqueId(value);
+    if (index < mCapacity)
+    {
+        /**
+         *  The data is stored in LittleEndian byte order in the set. This will enable
+         *  different machine architectures to interpret a given set in a consistent manner,
+         *  for serialize and deserialize operations.
+         */
+        mData[index] = ToLE64(value);
+        if (index == mNextAvailable)
+        {
+            mNextAvailable = static_cast<uint16_t>(index + 1);
+        }
+        err = CHIP_NO_ERROR;
+    }
+    return err;
+}
+
+void SerializableU64SetBase::Remove(uint64_t value)
+{
+    uint16_t index = Find(value);
+    if (index < mCapacity)
+    {
+        mData[index] = 0;
+        if ((index + 1) == mNextAvailable)
+        {
+            mNextAvailable = index;
+            while (mNextAvailable > 0 && mData[mNextAvailable - 1] == 0)
+                mNextAvailable--;
+        }
+    }
+}
+
+uint16_t SerializableU64SetBase::FirstAvailableForUniqueId(uint64_t value)
+{
+    /**
+     *  The data is stored in LittleEndian byte order in the set. This will enable
+     *  different machine architectures to interpret a given set in a consistent manner,
+     *  for serialize and deserialize operations.
+     */
+    uint64_t valueLE   = ToLE64(value);
+    uint16_t available = mNextAvailable;
+    for (uint16_t i = 0; i < mNextAvailable; i++)
+    {
+        if (mData[i] == valueLE)
+        {
+            return i;
+        }
+
+        if (mData[i] == 0 && i < available)
+        {
+            available = i;
```

Comment:
why not return i here? since i is increasing, "i < available" can only happen on the first iteration.

Reply:
just realized - cannot return because we need to search all entries for an existing data :( a comment would help confused people like myself.